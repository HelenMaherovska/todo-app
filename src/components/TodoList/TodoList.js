import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import TodoItemRow from "../TodoItemRow/TodoItemRow";
import { toggleTodo, deleteTodo, updateTodo } from "../../actions/todoActions";
import { getVisibleTodos } from "../../selectors/todoSelector";

export const TodoList = props => {
  const { todos, toggleTodo, deleteTodo, updateTodo } = props;

  if (!todos.length) {
    return <span className="empty-message">No todos in list</span>;
  }

  return (
    <table className="table">
      <thead>
        <tr>
          <th>Todo</th>
          <th>Date</th>
          <th />
        </tr>
      </thead>
      <tbody>
        {todos.map(todo => (
          <TodoItemRow
            key={todo.id}
            {...todo}
            onClick={toggleTodo}
            onDelete={deleteTodo}
            onUpdate={updateTodo}
          />
        ))}
      </tbody>
    </table>
  );
};

TodoList.propTypes = {
  todos: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      text: PropTypes.string.isRequired,
      completed: PropTypes.bool.isRequired,
      date: PropTypes.string.isRequired
    }).isRequired
  ).isRequired,
  toggleTodo: PropTypes.func.isRequired,
  deleteTodo: PropTypes.func.isRequired,
  updateTodo: PropTypes.func.isRequired
};

TodoList.defaultProps = {
  todos: []
};

const mapStateToProps = state => ({
  todos: getVisibleTodos(state)
});

const mapDispatchToProps = {
  toggleTodo,
  deleteTodo,
  updateTodo
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(TodoList);
