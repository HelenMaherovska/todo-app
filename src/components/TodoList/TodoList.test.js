import React from "react";
import { shallow } from "enzyme";
import { TodoList } from "./TodoList";

describe("TodoList", () => {
  const todos = [
    {
      id: "0",
      text: "todo item 1",
      completed: false
    },
    {
      id: "1",
      text: "todo item 2",
      completed: true
    },
    {
      id: "2",
      text: "todo item 3",
      completed: true
    }
  ];
  const mockToggle = jest.fn();
  const mockDelete = jest.fn();
  const mockUpdate = jest.fn();
  const props = {
    todos,
    toggleTodo: mockToggle,
    deleteTodo: mockDelete,
    updateTodo: mockUpdate
  };
  const todoList = shallow(<TodoList {...props} />);

  it("renders correctly", () => {
    expect(todoList).toMatchSnapshot();
  });
  it("renders a valid bootstrap table", () => {
    expect(todoList.find(".table")).toHaveLength(1);
  });

  describe("rendering list of todos", () => {
    it("creates a TodoItemRow component", () => {
      expect(todoList.find("TodoItemRow").exists()).toBe(true);
    });
  });
});
