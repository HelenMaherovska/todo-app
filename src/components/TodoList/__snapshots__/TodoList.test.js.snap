// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TodoList
    deleteTodo={[MockFunction]}
    todos={
      Array [
        Object {
          "completed": false,
          "id": "0",
          "text": "todo item 1",
        },
        Object {
          "completed": true,
          "id": "1",
          "text": "todo item 2",
        },
        Object {
          "completed": true,
          "id": "2",
          "text": "todo item 3",
        },
      ]
    }
    toggleTodo={[MockFunction]}
    updateTodo={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <thead>
          <tr>
            <th>
              Todo
            </th>
            <th>
              Date
            </th>
            <th />
          </tr>
        </thead>,
        <tbody>
          <TodoItemRow
            completed={false}
            id="0"
            onClick={[MockFunction]}
            onDelete={[MockFunction]}
            onUpdate={[MockFunction]}
            text="todo item 1"
          />
          <TodoItemRow
            completed={true}
            id="1"
            onClick={[MockFunction]}
            onDelete={[MockFunction]}
            onUpdate={[MockFunction]}
            text="todo item 2"
          />
          <TodoItemRow
            completed={true}
            id="2"
            onClick={[MockFunction]}
            onDelete={[MockFunction]}
            onUpdate={[MockFunction]}
            text="todo item 3"
          />
        </tbody>,
      ],
      "className": "table",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <tr>
            <th>
              Todo
            </th>
            <th>
              Date
            </th>
            <th />
          </tr>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <th>
                Todo
              </th>,
              <th>
                Date
              </th>,
              <th />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Todo",
              },
              "ref": null,
              "rendered": "Todo",
              "type": "th",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Date",
              },
              "ref": null,
              "rendered": "Date",
              "type": "th",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "th",
            },
          ],
          "type": "tr",
        },
        "type": "thead",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <TodoItemRow
              completed={false}
              id="0"
              onClick={[MockFunction]}
              onDelete={[MockFunction]}
              onUpdate={[MockFunction]}
              text="todo item 1"
            />,
            <TodoItemRow
              completed={true}
              id="1"
              onClick={[MockFunction]}
              onDelete={[MockFunction]}
              onUpdate={[MockFunction]}
              text="todo item 2"
            />,
            <TodoItemRow
              completed={true}
              id="2"
              onClick={[MockFunction]}
              onDelete={[MockFunction]}
              onUpdate={[MockFunction]}
              text="todo item 3"
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "class",
            "props": Object {
              "completed": false,
              "id": "0",
              "onClick": [MockFunction],
              "onDelete": [MockFunction],
              "onUpdate": [MockFunction],
              "text": "todo item 1",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "class",
            "props": Object {
              "completed": true,
              "id": "1",
              "onClick": [MockFunction],
              "onDelete": [MockFunction],
              "onUpdate": [MockFunction],
              "text": "todo item 2",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "2",
            "nodeType": "class",
            "props": Object {
              "completed": true,
              "id": "2",
              "onClick": [MockFunction],
              "onDelete": [MockFunction],
              "onUpdate": [MockFunction],
              "text": "todo item 3",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "tbody",
      },
    ],
    "type": "table",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <thead>
            <tr>
              <th>
                Todo
              </th>
              <th>
                Date
              </th>
              <th />
            </tr>
          </thead>,
          <tbody>
            <TodoItemRow
              completed={false}
              id="0"
              onClick={[MockFunction]}
              onDelete={[MockFunction]}
              onUpdate={[MockFunction]}
              text="todo item 1"
            />
            <TodoItemRow
              completed={true}
              id="1"
              onClick={[MockFunction]}
              onDelete={[MockFunction]}
              onUpdate={[MockFunction]}
              text="todo item 2"
            />
            <TodoItemRow
              completed={true}
              id="2"
              onClick={[MockFunction]}
              onDelete={[MockFunction]}
              onUpdate={[MockFunction]}
              text="todo item 3"
            />
          </tbody>,
        ],
        "className": "table",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <tr>
              <th>
                Todo
              </th>
              <th>
                Date
              </th>
              <th />
            </tr>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <th>
                  Todo
                </th>,
                <th>
                  Date
                </th>,
                <th />,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Todo",
                },
                "ref": null,
                "rendered": "Todo",
                "type": "th",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Date",
                },
                "ref": null,
                "rendered": "Date",
                "type": "th",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": "th",
              },
            ],
            "type": "tr",
          },
          "type": "thead",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <TodoItemRow
                completed={false}
                id="0"
                onClick={[MockFunction]}
                onDelete={[MockFunction]}
                onUpdate={[MockFunction]}
                text="todo item 1"
              />,
              <TodoItemRow
                completed={true}
                id="1"
                onClick={[MockFunction]}
                onDelete={[MockFunction]}
                onUpdate={[MockFunction]}
                text="todo item 2"
              />,
              <TodoItemRow
                completed={true}
                id="2"
                onClick={[MockFunction]}
                onDelete={[MockFunction]}
                onUpdate={[MockFunction]}
                text="todo item 3"
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "0",
              "nodeType": "class",
              "props": Object {
                "completed": false,
                "id": "0",
                "onClick": [MockFunction],
                "onDelete": [MockFunction],
                "onUpdate": [MockFunction],
                "text": "todo item 1",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "class",
              "props": Object {
                "completed": true,
                "id": "1",
                "onClick": [MockFunction],
                "onDelete": [MockFunction],
                "onUpdate": [MockFunction],
                "text": "todo item 2",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "2",
              "nodeType": "class",
              "props": Object {
                "completed": true,
                "id": "2",
                "onClick": [MockFunction],
                "onDelete": [MockFunction],
                "onUpdate": [MockFunction],
                "text": "todo item 3",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "tbody",
        },
      ],
      "type": "table",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
